### PR TITLE
Defining all Hypergrid types in gRPC

### PR DESCRIPTION
**Pull request by byte-sculptor**
The goal of this PR is to define all Hypergrid types in proto3 and build an encoder/decoder pair to support it. 

Those messages are not yet used by the gRPC service, as this would break the existing clients. A subsequent PR will implement encoders/decoders in C# at which point we will be able to switch the gRPC services to the new message types.

### CODE REVIEW

File: source/Mlos.Proto/OptimizerService.proto [hunk 1]
```diff
@@ -173,3 +173,109 @@ message SimpleString
...
+    repeated PrimitiveValue OrderedValues = 3;
+};
+
+message CategoricalDimension
+{
+    string Name = 1;
+    repeated PrimitiveValue Values = 2;
+};
+
+message CompositeDimension
```
Comment (grlap) [inline source/Mlos.Proto/OptimizerService.proto:214]: Comments

File: source/Mlos.Proto/OptimizerService.proto [hunk 2]
```diff
@@ -173,3 +173,109 @@ message SimpleString
...
+{
+    oneof Value {
+        int64 IntValue = 1;
+        double DoubleValue = 2;
+        bool BoolValue = 3;
+        string StringValue = 4;
+    }
+};
+
+message SimpleDimensionUnion
```
Comment (grlap) [inline source/Mlos.Proto/OptimizerService.proto:231]: Remove

File: source/Mlos.Proto/OptimizerService.proto [hunk 3]
```diff
@@ -173,3 +173,109 @@ message SimpleString
...
+{
+    oneof Dimension {
+        ContinuousDimension ContinuousDimension = 1;
+        DiscreteDimension DiscreteDimension = 2;
+        OrdinalDimension OrdinalDimension = 3;
+        CategoricalDimension CategoricalDimension = 4;
+    }
+};
+
+message AllDimensionUnion
```
Comment (grlap) [inline source/Mlos.Proto/OptimizerService.proto:241]: rename to Dimension

File: source/Mlos.Proto/OptimizerService.proto [hunk 4]
```diff
@@ -173,3 +173,109 @@ message SimpleString
...
+        ContinuousDimension ContinuousDimension = 1;
+        DiscreteDimension DiscreteDimension = 2;
+        OrdinalDimension OrdinalDimension = 3;
+        CategoricalDimension CategoricalDimension = 4;
+        EmptyDimension EmptyDimension = 5;
+        CompositeDimension CompositeDimension = 6;
+    }
+};
+
+message SimpleHypergrid
```
Comment (grlap) [inline source/Mlos.Proto/OptimizerService.proto:253]: as a next step, rename to Hypergrid

File: source/Mlos.Proto/OptimizerService.proto [hunk 5]
```diff
@@ -173,3 +173,109 @@ message SimpleString
...
+    repeated GuestSubgrid GuestSubgrids = 3;
+};
+
+message GuestSubgrid
+{
+    SimpleHypergrid Subgrid = 1;
+    AllDimensionUnion ExternalPivotDimension = 2;
+};
+
+message Point
```
Comment (grlap) [inline source/Mlos.Proto/OptimizerService.proto:266]: Comment

File: source/Mlos.Proto/OptimizerService.proto [hunk 6]
```diff
@@ -173,3 +173,109 @@ message SimpleString
 message Empty
 {
 };
+
+////// All Hypergrid Types
+
+message EmptyDimension
+{
+    string Name = 1;
+    string Type = 2;
```
Comment (grlap) [inline source/Mlos.Proto/OptimizerService.proto:182]: DimensionType

Comment (grlap) [reply source/Mlos.Proto/OptimizerService.proto:182]: Add an enum to describe dimension type

File: source/Mlos.Proto/OptimizerService.proto (hunk 1 above)
Comment (byte-sculptor) [reply source/Mlos.Proto/OptimizerService.proto:240]: Do we need more comments still?

**Review by grlap — APPROVED** [review]

**Pull request merged**